File: backend/app/routers/recipients.py
import uuid
from datetime import datetime, timezone

from fastapi import APIRouter, HTTPException, status

from app.mock_store import recipients
from app.schemas.recipient import RecipientCreate, RecipientRead, RecipientUpdate

router = APIRouter(prefix="/recipients", tags=["recipients"])


@router.get("/", response_model=list[RecipientRead])
async def list_recipients():
    return sorted(recipients.values(), key=lambda x: x["created_at"], reverse=True)


@router.post("/", response_model=RecipientRead, status_code=status.HTTP_201_CREATED)
async def create_recipient(body: RecipientCreate):
    rec = {"id": uuid.uuid4(), "created_at": datetime.now(timezone.utc), **body.model_dump()}
    recipients[rec["id"]] = rec
    return rec


@router.get("/{recipient_id}", response_model=RecipientRead)
async def get_recipient(recipient_id: uuid.UUID):
    rec = recipients.get(recipient_id)
    if not rec:
        raise HTTPException(status_code=404, detail="Recipient not found")
    return rec


@router.put("/{recipient_id}", response_model=RecipientRead)
async def update_recipient(recipient_id: uuid.UUID, body: RecipientUpdate):
    rec = recipients.get(recipient_id)
    if not rec:
        raise HTTPException(status_code=404, detail="Recipient not found")
    rec.update(body.model_dump(exclude_unset=True))
    return rec


@router.delete("/{recipient_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_recipient(recipient_id: uuid.UUID):
    if recipient_id not in recipients:
        raise HTTPException(status_code=404, detail="Recipient not found")
    del recipients[recipient_id]

File: backend/src/routers/recipients.js
'use strict';

const { Router } = require('express');
const { v4: uuidv4 } = require('uuid');
const { recipients } = require('../mockStore');

const router = Router();

// GET /api/recipients
router.get('/', (_req, res) => {
  const items = [...recipients.values()].sort((a, b) => b.created_at.localeCompare(a.created_at));
  res.json(items);
});

// POST /api/recipients
router.post('/', (req, res) => {
  const { email, name = null, is_active = true } = req.body;
  if (!email || !email.includes('@')) {
    return res.status(422).json({ detail: 'valid email is required' });
  }
  const rec = { id: uuidv4(), email, name, is_active, created_at: new Date().toISOString() };
  recipients.set(rec.id, rec);
  res.status(201).json(rec);
});

// GET /api/recipients/:id
router.get('/:id', (req, res) => {
  const rec = recipients.get(req.params.id);
  if (!rec) return res.status(404).json({ detail: 'Recipient not found' });
  res.json(rec);
});

// PUT /api/recipients/:id
router.put('/:id', (req, res) => {
  const rec = recipients.get(req.params.id);
  if (!rec) return res.status(404).json({ detail: 'Recipient not found' });
  const allowed = ['email', 'name', 'is_active'];
  for (const key of allowed) {
    if (req.body[key] !== undefined) rec[key] = req.body[key];
  }
  res.json(rec);
});

// DELETE /api/recipients/:id
router.delete('/:id', (req, res) => {
  if (!recipients.has(req.params.id)) return res.status(404).json({ detail: 'Recipient not found' });
  recipients.delete(req.params.id);
  res.status(204).end();
});

module.exports = router;
